### PR TITLE
Metadata: only include relevant files in the source list

### LINKED ISSUE #2561: Determine which source files are necessary to compile a certain contract and only specify those in the metadata
Fixes https://github.com/ethereum/solidity/issues/2102

### PR DESCRIPTION
**Pull request by axic**
Fixes #2561.

### CODE REVIEW

**Comment by axic** [comment]
This works for me, but the cyclic import check is a mess at the moment.

`_2561.sol`:
```
pragma solidity ^0.4.0;

import "_2561_2.sol";

contract C {
}
```

`_2561_2.sol`:
```
pragma solidity ^0.4.0;

import "_2561_3.sol";
```

`_2561_3.sol`:
```
pragma solidity ^0.4.0;

import "_2561_2.sol";
import "_2561_4.sol";
```

**Comment by chriseth** [comment]
Sorry, @VoR0220 it seems @axic was faster...

File: libsolidity/ast/AST.cpp [hunk 1]
```diff
@@ -84,13 +84,41 @@ SourceUnitAnnotation& SourceUnit::annotation() const
 	return dynamic_cast<SourceUnitAnnotation&>(*m_annotation);
 }
 
-string Declaration::sourceUnitName() const
+std::set<SourceUnit const*> SourceUnit::referencedSourceUnits(bool _flatten, std::set<SourceUnit const*> _skipList) const
+{
+	std::set<SourceUnit const*> sourceUnits;
+	for (auto const& node: nodes())
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: You could use `for (ImportDirective const* importDirective: filteredNodes<ImportDirective>(nodes())) { ... }`

**Comment by axic** [comment]
Sorry didn't know @VoR0220 picked it up. @VoR0220, do you want to clean up the cyclic import part? :)

File: libsolidity/ast/AST.cpp [hunk 2]
```diff
@@ -84,13 +84,41 @@ SourceUnitAnnotation& SourceUnit::annotation() const
...
+	{
+		if (auto const* importDirective = dynamic_cast<ImportDirective const*>(node.get()))
+		{
+			auto annotation = importDirective->annotation();
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: Please use `auto const&` here, otherwise it performs a copy.

File: libsolidity/ast/AST.cpp [hunk 3]
```diff
@@ -84,13 +84,41 @@ SourceUnitAnnotation& SourceUnit::annotation() const
 	return dynamic_cast<SourceUnitAnnotation&>(*m_annotation);
 }
 
-string Declaration::sourceUnitName() const
+std::set<SourceUnit const*> SourceUnit::referencedSourceUnits(bool _flatten, std::set<SourceUnit const*> _skipList) const
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: No need for `std::` prefixes here.

File: libsolidity/ast/AST.h [hunk 4]
```diff
@@ -131,6 +131,9 @@ class SourceUnit: public ASTNode
 
 	std::vector<ASTPointer<ASTNode>> nodes() const { return m_nodes; }
 
+	/// @returns a set of referenced SourceUnits. Flattens them if @a _flatten is true.
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: Please explain the term "flatten" and also explain the `_skipList` parameter.

Comment (chriseth) [reply libsolidity/ast/AST.h]: Shouldn't `flatten` be rather called `recurse`?

File: libsolidity/ast/AST.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/ast/AST.cpp]: Perhaps it is better not to expose the `_skipList` in the public interface.

File: libsolidity/ast/AST.cpp [hunk 5]
```diff
@@ -84,13 +84,41 @@ SourceUnitAnnotation& SourceUnit::annotation() const
...
+{
+	std::set<SourceUnit const*> sourceUnits;
+	for (auto const& node: nodes())
+	{
+		if (auto const* importDirective = dynamic_cast<ImportDirective const*>(node.get()))
+		{
+			auto annotation = importDirective->annotation();
+			sourceUnits.insert(annotation.sourceUnit);
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: Can`t you move that into the if body?

File: libsolidity/ast/AST.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/ast/AST.cpp]: We actually only use the `sourceUnit` from the annotation, so we could store this instead in a local variable.

File: libsolidity/ast/AST.cpp [hunk 6]
```diff
@@ -84,13 +84,41 @@ SourceUnitAnnotation& SourceUnit::annotation() const
...
+			if (!_skipList.count(annotation.sourceUnit))
+			{
+				_skipList.insert(annotation.sourceUnit);
+				if (_flatten)
+				{
+					std::set<SourceUnit const*> referencedSourceUnits = annotation.sourceUnit->referencedSourceUnits(true, _skipList);
+					sourceUnits.insert(referencedSourceUnits.begin(), referencedSourceUnits.end());
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: `+=` cannot be used here?

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by chriseth** [comment]
What is the problem with cyclic imports?

**Comment by axic** [comment]
The messy `skipList` parameter.

**Comment by chriseth** [comment]
We use that all the time. Perhaps some generic template solution for that would be nice...

File: libsolidity/ast/AST.cpp (hunk 6 above)
Comment (axic) [reply libsolidity/ast/AST.cpp]: Not sure, this was the suggested way by SE 😉

File: libsolidity/ast/AST.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/ast/AST.cpp]: Removed `std::` prefixes.

File: libsolidity/ast/AST.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/ast/AST.cpp]: Changed.

File: libsolidity/ast/AST.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/ast/AST.cpp]: Changed.

File: libsolidity/ast/AST.h (hunk 4 above)
Comment (axic) [reply libsolidity/ast/AST.h]: Used flatten from "flattened arrays" used by Ruby/Python, but I guess recurse could go too.

File: libsolidity/ast/AST.cpp (hunk 6 above)
Comment (axic) [reply libsolidity/ast/AST.cpp]: There's no such overload. The template in `CommonData` is only for vectors.

File: libsolidity/ast/AST.h (hunk 4 above)
Comment (chriseth) [reply libsolidity/ast/AST.h]: Sure, but if flatten is false, it does not return a non-flattened structure, so I think recurse is a much better fit.

**Comment by chriseth** [comment]
Rename and good to go.

**Comment by chriseth** [comment]
Oh and some tests :-)

**Comment by axic** [comment]
Renamed `flatten` and added tests.

File: libsolidity/ast/AST.cpp [hunk 7]
```diff
@@ -84,13 +84,35 @@ SourceUnitAnnotation& SourceUnit::annotation() const
 	return dynamic_cast<SourceUnitAnnotation&>(*m_annotation);
 }
 
-string Declaration::sourceUnitName() const
+set<SourceUnit const*> SourceUnit::referencedSourceUnits(bool _recurse, set<SourceUnit const*> _skipList) const
```
Comment (axic) [inline libsolidity/ast/AST.cpp:87]: `skipList` could more accurately be called `visitedList` (while `sourceUnits` contains imported files not yet visited too)

**Comment by axic** [comment]
@chriseth can you review again? Should be ready (see last comment for a possible improvement).

**Comment by axic** [comment]
Need to rebase after #2599.

**Review by chriseth — APPROVED** [review]
This has quite bad performance, but I guess we will see whether it is acutally relevant.

**Pull request merged**